community/cloud-foundation/templates/cloud_spanner/tests/integration/cloud_spanner.bats

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env bats
+
+source tests/helpers.bash
+
+TEST_NAME=$(basename "${BATS_TEST_FILENAME}" | cut -d '.' -f 1)
+
+## Create a random 10-char string and save it in a file.
+RANDOM_FILE="/tmp/${CLOUD_FOUNDATION_ORGANIZATION_ID}-${TEST_NAME}.txt"
+if [[ ! -e "${RANDOM_FILE}" ]]; then
+    RAND=$(head /dev/urandom | LC_ALL=C tr -dc a-z0-9 | head -c 10)
+    echo ${RAND} > "${RANDOM_FILE}"
+fi
+
+# Set variables based on the random string saved in the file.
+# envsubst requires all variables used in the example/config to be exported.
+if [[ -e "${RANDOM_FILE}" ]]; then
+    export RAND=$(cat "${RANDOM_FILE}")
+    DEPLOYMENT_NAME="${CLOUD_FOUNDATION_PROJECT_ID}-${TEST_NAME}-${RAND}"
+    # Replace underscores in the deployment name with dashes.
+    DEPLOYMENT_NAME=${DEPLOYMENT_NAME//_/-}
+    CONFIG=".${DEPLOYMENT_NAME}.yaml"
+fi
+
+export PROJECT_NUMBER=$(gcloud projects list | grep dev-island | awk '{ print $3}')
+
+########## HELPER FUNCTIONS ##########
+
+function create_config() {
+    echo "Creating ${CONFIG}"
+    envsubst < "templates/cloud_spanner/tests/integration/${TEST_NAME}.yaml" > "${CONFIG}"
+}
+
+function delete_config() {
+    echo "Deleting ${CONFIG}"
+    rm -f "${CONFIG}"
+}
+
+function setup() {
+    if [ ${BATS_TEST_NUMBER} -eq 1 ]; then
+        create_config
+    fi
+}
+
+function teardown() {
+    if [[ "$BATS_TEST_NUMBER" -eq "${#BATS_TEST_NAMES[@]}" ]]; then
+        rm -f "${RANDOM_FILE}"
+        delete_config
+    fi
+}
+
+@test "Creating deployment ${DEPLOYMENT_NAME} from ${CONFIG}" {
+    run gcloud deployment-manager deployments create "${DEPLOYMENT_NAME}" \
+        --config ${CONFIG} \
+        --project "${CLOUD_FOUNDATION_PROJECT_ID}"
+    [[ "$status" -eq 0 ]]
+}
```

Comment:
I'm getting the following error when running the tests:

```
 ✗ Creating deployment ocsi-courseralab1-guestbook-cloud-spanner-77y078y02m from .ocsi-courseralab1-guestbook-cloud-spanner-77y078y02m.yaml
   (in test file templates/cloud_spanner/tests/integration/cloud_spanner.bats, line 55)
     `[[ "$status" -eq 0 ]]' failed
```

I don't see where `status` is being set; did you mean `$?` ?

Reply:
I verified the test and they work for me.

Looks like the deployment failed and according to the DM logs, specifically the setting spanner IAM. Can you check if you have `roles/spanner.admin` permissions?

Reply:
Ok, I investigated some more and it looks like the `${PROJECT_NUMBER}` is empty resulting in the following YAML file:

```yaml
 imports:
   - path: templates/cloud_spanner/cloud_spanner.py
     name: cloud_spanner.py
   
   resources:
     - name: test-myspannercluster-5simlkgtbz
       type: cloud_spanner.py
       properties:
         displayName: "Spanner Cluster 1"
         nodeCount: 2
         instanceConfig: nam3
         bindings:
           - role: "roles/spanner.admin"
             members:
               - "serviceAccount:@cloudservices.gserviceaccount.com"
         databases:
           - name: "spannerdb1"
             bindings:
               - role: "roles/spanner.databaseAdmin"
                 members:
                   - "serviceAccount:@cloudservices.gserviceaccount.com"
```

It makes sense given the DM error message:

```shell
..."ResourceErrorCode":"400","ResourceErrorMessage":{"code":400,"message":"Invalid service account email (@cloudservices.gserviceaccount.com).","status":"INVALID_ARGUMENT","statusMessage":"Bad
    Request",...
```

Is there any config I need to add to run this template?

Reply:
Found the problem. Please try with the latest commit.

Reply:
Hi Vince, not yet - still failing. This is what I get now:

```shell
The fingerprint of the deployment is SepRNUJAbtV-1y96ebgZrA==
Waiting for create [operation-1540936892736-57979541ce200-199f3a06-b818546e]...
.......................failed.
ERROR: (gcloud.deployment-manager.deployments.create) Error in Operation [operation-1540936892736-57979541ce200-199f3a06-b818546e]: errors:
- code: RESOURCE_ERROR
  location: /deployments/ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a/resources/test-myspannercluster-16g41lij4a-setIamPolicy
  message: '{"ResourceType":"gcp-types/spanner-v1:spanner.projects.instances.setIamPolicy","ResourceErrorCode":"400","ResourceErrorMessage":{"code":400,"message":"Service
    account creator@cloudservices.gserviceaccount.com does not exist.","status":"INVALID_ARGUMENT","statusMessage":"Bad
    Request","requestPath":"https://spanner.googleapis.com/v1/projects/ocsi-courseralab1-guestbook/instances/test-myspannercluster-16g41lij4a:setIamPolicy","httpMethod":"POST"}}'
```

Generated YAML looks like this:

```yaml
imports:
   - path: templates/cloud_spanner/cloud_spanner.py
     name: cloud_spanner.py
   
   resources:
     - name: test-myspannercluster-16g41lij4a
       type: cloud_spanner.py
       properties:
         displayName: "Spanner Cluster 1"
         nodeCount: 2
         instanceConfig: nam3
         bindings:
           - role: "roles/spanner.admin"
             members:
               - "serviceAccount:creator@cloudservices.gserviceaccount.com"
         databases:
           - name: "spannerdb1"
             bindings:
               - role: "roles/spanner.databaseAdmin"
                 members:
                   - "serviceAccount:creator@cloudservices.gserviceaccount.com"
```

Reply:
PS: to be more specific these are the test results:
```shell
 ✓ Verifying that Spanner cluster was created as part of ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a
 ✗ Verifying that Spanner cluster IAM was created as part of ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a
   (in test file templates/cloud_spanner/tests/integration/cloud_spanner.bats, line 70)
     `[[ "$output" =~ "${PROJECT_NUMBER}@cloudservices.gserviceaccount.com" ]]' failed
 ✓ Verifying that Spanner DB was created as part of ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a
 ✗ Verifying that Spanner DB IAM was created as part of ocsi-courseralab1-guestbook-cloud-spanner-16g41lij4a
   (in test file templates/cloud_spanner/tests/integration/cloud_spanner.bats, line 84)
     `[[ "$output" =~ "role: roles/spanner.databaseAdmin" ]]' failed
 ✓ Deleting deployment
```